java-vtl-script/src/main/java/no/ssb/vtl/script/functions/VTLSubstr.java

```diff
@@ -0,0 +1,98 @@
+ */
+
+import no.ssb.vtl.model.VTLInteger;
+import no.ssb.vtl.model.VTLString;
+
+import static java.lang.String.*;
+
+public class VTLSubstr extends AbstractVTLFunction<VTLString> {
+
+    private static final String ARGUMENT_GREATER_THAT_ZERO = "%s must be greater than zero, was %s";
+    private static final Argument<VTLString> DS = new Argument<>("ds", VTLString.class);
+    private static final Argument<VTLInteger> START_POSITION = new Argument<>("startPosition", VTLInteger.class);
+    private static final Argument<VTLInteger> LENGTH = new Argument<>("length", VTLInteger.class);
+    private static VTLSubstr instance;
+    private VTLString STRING_NULL = VTLString.of((String) null);
+    private VTLInteger INTEGER_NULL = VTLInteger.of((Long) null);
+
+    private VTLSubstr() {
+        super("substr", VTLString.class, DS, START_POSITION, LENGTH);
+    }
+
+    public static VTLSubstr getInstance() {
+        if (instance == null)
+            instance = new VTLSubstr();
+        return instance;
+    }
+
+    @Override
+    protected VTLString safeInvoke(TypeSafeArguments arguments) {
+        VTLString value = arguments.getNullable(DS, STRING_NULL);
+        VTLInteger startPosition = arguments.getNullable(START_POSITION, INTEGER_NULL);
+        VTLInteger length = arguments.getNullable(LENGTH, INTEGER_NULL);
+
+        if (value.get() == null) {
+            return VTLString.of((String) null);
+        }
+
+        if (value.get().isEmpty()) {
+            return VTLString.of("");
+        }
+
+        if (startPosition.get() == null) {
+            return VTLString.of(value.get());
+        }
+
+        if (startPosition.get().intValue() < 0) {
+            throw new IllegalArgumentException(
+                    format(ARGUMENT_GREATER_THAT_ZERO, START_POSITION, startPosition)
+            );
+        }
+
+        if (length.get() != null && length.get().intValue() < 0) {
+            throw new IllegalArgumentException(
+                    format(ARGUMENT_GREATER_THAT_ZERO, LENGTH, length)
+            );
+        }
+
+        int substringLength = value.get().length() - 1 - startPosition.get().intValue();
+        if (substringLength < 0) {
+            //deviation from the VTL specification 1.1
+            return VTLString.of("");
+        }
+
+        if (ignoreLengthArgument(startPosition.get(), length.get(), value.get())) {
+            return VTLString.of(value.get().substring(startPosition.get().intValue()));
+        } else {
+            Long endIndex = startPosition.get() + length.get();
+            return VTLString.of(value.get().substring(startPosition.get().intValue(), endIndex.intValue()));
+        }
+    }
+
+    private boolean ignoreLengthArgument(Long startPosition, Long length, String value) {
+        if (length == null) {
+            return true;
+        }
+
+        Long endIndex = startPosition + length - 1;
+        return endIndex > value.length() - 1;
```

Comment:
Could you not simply return a valid `endIndex` instead of calculating the index and the check if it is valid, and then if it is valid it is recalculated in the else block
Also: Double `-1` is unnecessary.